src/Raven.Server/Rachis/ClusterMaintenanceMaster.cs

```diff
@@ -0,0 +1,317 @@
+                                    startUpdateTime = DateTime.UtcNow;
+                                    var readResponseTask = context.ReadForMemoryAsync(connection, _readStatusUpdateDebugString, _token);
+                                    if (await Task.WhenAny(readResponseTask, timeout) == timeout)
+                                    {
+                                        //TODO : logging about timeout
+                                        ReceivedReport = new ClusterNodeStatusReport(new Dictionary<string, DatabaseStatusReport>(), 
+                                            ClusterNodeStatusReport.ReportStatus.Timeout,
+                                            null,
+                                            startUpdateTime,
+                                            DateTime.UtcNow);
+                                        await readResponseTask;
+                                        needToWait = true;
+                                        break;
+                                    }
+
+                                    using (var statusUpdateJson = await readResponseTask)
+                                    {
+                                        var report = new Dictionary<string, DatabaseStatusReport>();
+                                        foreach (var property in statusUpdateJson.GetPropertyNames())
+                                        {
+                                            var value = (BlittableJsonReaderObject)statusUpdateJson[property];
+                                            report.Add(property, JsonDeserializationServer.DatabaseStatusReport(value));
+                                        }
+                                        ReceivedReport = new ClusterNodeStatusReport(
+                                            report,
+                                            ClusterNodeStatusReport.ReportStatus.Ok,
+                                            null,
+                                            startUpdateTime,
+                                            DateTime.UtcNow);
+                                    }
+                                }
+                            }
+
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        //TODO : do not forget to add this to log as well
+                        ReceivedReport = new ClusterNodeStatusReport(new Dictionary<string, DatabaseStatusReport>(), 
+                            ClusterNodeStatusReport.ReportStatus.Error,
+                            e,
+                            startUpdateTime,
+                            DateTime.UtcNow);
+                        needToWait = true;
+                    }
+                }
+            }
+
+            private async Task<Stream> ConnectAndGetNetworkStreamAsync(TcpConnectionInfo tcpConnectionInfo)
```

Comment:
Why do we have this method here, shouldn't this be in `TcpUtils` and have that as the only way to create a connection?